include/iceflow/producer.hpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2024 The IceFlow Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ICEFLOW_PRODUCER_HPP
+#define ICEFLOW_PRODUCER_HPP
+
+#include "iceflow.hpp"
+
+namespace iceflow {
+
+/**
+ * Allows for publishing data to `IceflowConsumer`s.
+ *
+ * The data is split across one or more `topicPartitions` whose indexes can be
+ * passed as a constructor argument.
+ */
+class IceflowProducer {
+public:
+  IceflowProducer(std::shared_ptr<IceFlow> iceflow, const std::string &pubTopic,
+                  const std::vector<int> &topicPartitions,
+                  std::chrono::milliseconds publishInterval)
+      : m_iceflow(iceflow), m_pubTopic(pubTopic),
+        m_topicPartitions(topicPartitions), m_publishInterval(publishInterval),
+        m_lastPublishTimePoint(std::chrono::steady_clock::now()) {
+
+    if (auto validIceflow = m_iceflow.lock()) {
+      std::function<QueueEntry(void)> popQueueValueCallback =
+          std::bind(&IceflowProducer::popQueueValue, this);
+      std::function<bool(void)> hasQueueValueCallback =
+          std::bind(&IceflowProducer::hasQueueValue, this);
+      std::function<std::chrono::nanoseconds(void)>
+          getTimeUntilNextPublishCallback =
+              std::bind(&IceflowProducer::getTimeUntilNextPublish, this);
+
+      ProducerRegistrationInfo producerRegistration = {
+          popQueueValueCallback,
+          hasQueueValueCallback,
+          getTimeUntilNextPublishCallback,
+      };
+
+      m_subscriberId = validIceflow->registerProducer(producerRegistration);
+    } else {
+      throw std::runtime_error("Iceflow instance has already expired.");
+    }
+  }
+
+  ~IceflowProducer() {
+    if (auto validIceflow = m_iceflow.lock()) {
+      validIceflow->deregisterProducer(m_subscriberId);
+    }
+  }
+
+  void pushData(const std::vector<uint8_t> &data) { m_outputQueue.push(data); }
+
+private:
+  QueueEntry popQueueValue() {
+    int partitionIndex = m_partitionCount++ % m_topicPartitions.size();
+    auto partitionNumber = m_topicPartitions[partitionIndex];
+    auto data = m_outputQueue.waitAndPopValue();
+    m_lastPublishTimePoint = std::chrono::steady_clock::now();
```

Comment:
Identifying the time of the last publish based on when `popQueueValue()` was called might lead to the actual publish interval being shorter than the defined publish interval if the thread is paused after retrieving the queue element, but before actually sending the data.

This _might_ be fine, depending on whether we consider the publish interval a strict minimum or only a recommendation by the user.

If it should be a strict minimum, we should define a callback for the producer that is called after actually publishing the data, which then updates the last publish time.

Reply:
> If it should be a strict minimum, we should define a callback for the producer that is called after actually publishing the data, which then updates the last publish time.

Yeah, that definitely makes sense, I will update the PR accordingly soon :) I also now remembered another slight change I wanted to make in this part of the code which I will add in the same wake. I will ping you once again as soon as these changes are implemented :)

Reply:
I hope I now improved the implementation in the latest commit – looking forward to your feedback on Monday :)